### PR TITLE
Add deprecation warning handlers

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title

### CODE REVIEW

File: numba/core/errors.py [hunk 1]
```diff
@@ -392,6 +393,34 @@ def wrapper(*args, **kwargs):
         return decorator
 
 
+def deprecate_moved_module(old_module, new_module, stacklevel=3):
+    """Warn about a module level location move of some part of Numba's
+    internals. stacklevel is 3 by default as most warning locations are
+    from `numba.XYZ` shims.
+    """
+
+    msg = ("An import of was requested from a module that has moved "
```
Comment (eric-wieser) [inline numba/core/errors.py:402]:
```suggestion
    msg = ("An import was requested from a module that has moved "
```

File: numba/core/errors.py [hunk 2]
```diff
@@ -392,6 +393,34 @@ def wrapper(*args, **kwargs):
...
+    locations
+    """
+    def wrapper(attr):
+        import_module(new_module)
+        msg = ("An import of was requested from a module that has moved "
+               "location.\nImport requested from: '{}', please update to use "
+               "'{}' or pin to Numba version 0.48.0.")
+        warnings.warn(msg.format(old_module, new_module),
+                        category=NumbaDeprecationWarning, stacklevel=2)
+        return getattr(new_module, attr)
```
Comment (eric-wieser) [inline numba/core/errors.py:420]:
```suggestion
        new_module_obj = import_module(new_module)
        msg = ("An import of was requested from a module that has moved "
               "location.\nImport requested from: '{}', please update to use "
               "'{}' or pin to Numba version 0.48.0.")
        warnings.warn(msg.format(old_module, new_module),
                        category=NumbaDeprecationWarning, stacklevel=2)
        return getattr(new_module_obj, attr)
```

I wonder if you want to print the deprecation if the `getattr` fails? Might make sense to swap it around.

File: numba/core/errors.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/core/errors.py:402]: Thanks, will patch.

File: numba/core/errors.py (hunk 2 above)
Comment (stuartarchibald) [reply numba/core/errors.py:420]: Excellent point, hadn't considered this, will swap them.

**Comment by stuartarchibald** [comment]
Thanks for taking a look @eric-wieser, I'll incorporate these suggestions into https://github.com/numba/numba/pull/5352 when I merge the patch in.

**Pull request closed without merging**